# Class 11: Derivatives 

Ch. 5, sections 10 - 12: Derivatives, Second derivatives, & Interpolation    

- Complete the activities as instructed by the professor

#### Import any packages we need below (update as we go):

In [1]:
import numpy as np
import math

### Forward, Backward, and Central difference 

Evaluate the derivative of $f(x) = e^{\sin(2x)}$ at $x = 0.5$

If you took the derivative by hand, what would it be? 

In [2]:
def fprime(x):
    return 2*np.cos(2*x)*np.exp(np.sin(2*x))

x = 0.5
true_val =fprime(x)
print(true_val)

2.506761534986894


Now apply forward/backward difference 

In [ ]:
def f(x):
    return np.exp(np.sin(2*x))

#apply forward difference
def calc_fd(f,x,h):
    fd = (f(x+h)-f(x))/h
    return fd

#backward difference
def calc_bd(f,x,h):
    fd = (f(x)-f(x-h))/h
    return fd
#this can be called and used just like the fd func


x= 0.5
h_fd = [10**(-i) for i in range(1,12)]

fd_array = np.zeros(len(h_fd))

for j in range(len(fd_array)):
    fd_array[j]= calc_fd(f,x,h_fd[j])


fd_error = np.abs(fd_array - true_val)

print("the true value is ", true_val)
print(" h      forward diff        error")
rowf = "{0:1.0e} {1:1.16f} {2:1.16f}"
for i in range(len(h_fd)):
    print(rowf.format(h_fd[i], fd_array[i], fd_error[i]))

the true value is  2.506761534986894
 h      forward diff        error
1e-01 2.1990570766492690 0.3077044583376249
1e-02 2.4807256192968197 0.0260359156900742
1e-03 2.5042064928371133 0.0025550421497806
1e-04 2.5065065168927703 0.0002550180941236
1e-05 2.5067360380326420 0.0000254969542519
1e-06 2.5067589857208361 0.0000025492660578
1e-07 2.5067612785534266 0.0000002564334673
1e-08 2.5067615094798157 0.0000000255070782
1e-09 2.5067619091601046 0.0000003741732106
1e-10 2.5067636855169440 0.0000021505300500
1e-11 2.5067947717616335 0.0000332367747395


Now apply central difference 

In [7]:

def calc_cd(f,x,h):
    cd = (f(x+h/2)-f(x-h/2))/h
    return cd

cd_array = np.zeros(len(h_fd))

for j in range(len(cd_array)):
    cd_array[j]= calc_cd(f,x,h_fd[j])

cd_error = np.abs(cd_array - true_val)

print("the true value is ", true_val)
print(" h      central diff        error")
rowf = "{0:1.0e} {1:1.16f} {2:1.16f}"
for i in range(len(h_fd)):
    print(rowf.format(h_fd[i], cd_array[i], cd_error[i]))

the true value is  2.506761534986894
 h      central diff        error
1e-01 2.4932959255171205 0.0134656094697734
1e-02 2.5066264877375843 0.0001350472493096
1e-03 2.5067601844752652 0.0000013505116288
1e-04 2.5067615214791061 0.0000000135077878
1e-05 2.5067615348373096 0.0000000001495843
1e-06 2.5067615352369899 0.0000000002500959
1e-07 2.5067615361251683 0.0000000011382744
1e-08 2.5067615538887367 0.0000000189018428
1e-09 2.5067614650708947 0.0000000699159992
1e-10 2.5067636855169440 0.0000021505300500
1e-11 2.5067503628406484 0.0000111721462455


How do they compare? 

### Second derivatives

Let's take the second derivative use central difference of the same function $f(x) = e^{\sin(2x)}$ at $x = 0.5$ & compare with the true value 

$$ \frac{d^2}{dx^2}e^{\sin(2x)} = 2 e^{\sin(2x)}(-2\sin(2x)+\cos(4x)+1)  $$

In [ ]:
def fdoubleprime(x):
    return 2*np.exp(np.sin(2*x))*(-2*np.sin(2*x)+np.cos(4*x)+1)

def second_deriv(f,x,h):
    second_d = (f(x+h) - 2*f(x)+f(x-h))/h**2
    return second_d

x = 0.5
true_val_doub = fdoubleprime(x)

second_array = np.zeros(len(h_fd))

for j in range(len(second_array)):
    second_array[j]= fdoubleprime(second_deriv,x,h_fd[j])


double_error = np.abs(second_array - true_val_doub)

print("the true value is ", true_val_doub)
print(" h      second diff        error")
rowf = "{0:1.0e} {1:1.16f} {2:1.16f}"
for i in range(len(h_fd)):
    print(rowf.format(h_fd[i], second_array[i], double_error[i]))


TypeError: fdoubleprime() takes 1 positional argument but 3 were given

### Automatic Differentiation - book example

Store value/derivative pairs as two-element lists in Python of the form [value, derivative]

Define some functions to perform basic mathematics operations on these inputted lists. 

Apply this to overall figure out the value of the function $f(x) = 3x^2 +e^x$ and its derivative at $t = 1/2$

To do this, perform the following sequence of operations: 

$$ u_1 = x^2$$
$$ u_2 = 3u_1$$
$$ u_3 = e^x$$

and finally 

$$ f(x) = u_2 + u_3$$

Each each mathematical operation will be performed using the generalized operations listed above and the end result will be a pair $[f,f']$, representing the value of $f$ and its derivative.

Final bit of info: to calculate the derivative $f'(t)$ at any value of $t$, need an input list to evaluate our function $f(𝑥)$. Use $[t,1]$ since for whatever value of $t$ we are interested in, by definition, this is the correct value/derivative pair for the independent variable $t$, because $dt/dt = 1$ always. So if we use this pair as the argument of our function, we will get out a pair representing the value of the function $f(t)$ and its derivative $f'(t)$, at the chosen value of $t$.


In [11]:
def power(x,c):
    vx, dx = x
    return [vx**c, c*vx**(c-1)*dx]
            
def mult(x,c):
    vx,dx = x
    return [c*vx,c+dx]


def expon(x):
    vx,dx = x
    return [np.exp(vx), np.exp(vx)*dx]

def add(x,y):
    vx,dx = x
    vy,dy = y
    return [vx+vy, dx+dy]


def f_ad(x):
    u1 = power(x,2)
    u2 = mult(u1,3)
    u3 = expon(x)
    u4 = add(u2,u3)

    return u4


print(f_ad([0.5,1]))



[np.float64(2.398721270700128), np.float64(5.648721270700128)]


The actual value of the function and the derivative at $f(t)$ at $t = 1/2$ is 
$$f (t) = 3t^2 + e^t = \frac{3}{4} + e^{1/2} = 2.3987 ...$$
$$ f'(t) = 6t + e^t = 3 + e^{1/2} = 4.6487 ...$$

Look how exact we are!